components/org.wso2.carbon.identity.application.authenticator.push/src/main/java/org/wso2/carbon/identity/application/authenticator/push/PushAuthenticator.java

```diff
@@ -186,19 +197,32 @@ protected void processAuthenticationResponse(HttpServletRequest httpServletReque
             throw new AuthenticationFailedException(errorMessage, e);
         }
         if (claimsSet != null) {
-            validateChallenge(claimsSet, serverChallenge, deviceId);
-
-            String authStatus =
-                    getClaimFromClaimSet(claimsSet, PushAuthenticatorConstants.TOKEN_RESPONSE, deviceId);
+            if (validator.validateChallenge(claimsSet, serverChallenge, deviceId)) {
+                String authStatus;
+                try {
+                    authStatus =
+                            validator.getClaimFromClaimSet(claimsSet, PushAuthenticatorConstants.TOKEN_RESPONSE,
+                                    deviceId);
+                } catch (PushAuthTokenValidationException e) {
+                    String errorMessage = String.format("Error in getting claim %s from the auth response token" +
+                            " received from device: %s", PushAuthenticatorConstants.TOKEN_SESSION_DATA_KEY, deviceId);
+                    throw new AuthenticationFailedException(errorMessage, e);
+                }
 
-            if (authStatus.equals(PushAuthenticatorConstants.AUTH_REQUEST_STATUS_SUCCESS)) {
-                authenticationContext.setSubject(user);
-            } else if (authStatus.equals(PushAuthenticatorConstants.AUTH_REQUEST_STATUS_DENIED)) {
-                redirectRetryPage(httpServletResponse, PushAuthenticatorConstants.AUTH_DENIED_PARAM,
-                        PushAuthenticatorConstants.AUTH_DENIED_MESSAGE, user);
+                if (authStatus.equals(PushAuthenticatorConstants.AUTH_REQUEST_STATUS_SUCCESS)) {
+                    authenticationContext.setSubject(user);
+                } else if (authStatus.equals(PushAuthenticatorConstants.AUTH_REQUEST_STATUS_DENIED)) {
+                    redirectRetryPage(httpServletResponse, PushAuthenticatorConstants.AUTH_DENIED_PARAM,
+                            PushAuthenticatorConstants.AUTH_DENIED_MESSAGE, user);
+                } else {
+                    String errorMessage = String.format("Authentication failed! Auth status for user" +
+                            " '%s' is not available in JWT.", user.toFullQualifiedUsername());
+                    throw new AuthenticationFailedException(errorMessage);
+                }
             } else {
-                String errorMessage = String.format("Authentication failed! Auth status for user" +
-                        " '%s' is not available in JWT.", user.toFullQualifiedUsername());
+                String errorMessage = String
+                        .format("Authentication failed! JWT challenge validation for device: %s of user: %s.",
+                                deviceId, user);
```

Comment:
User object is passed and not a string

Reply:
@pulasthi7 wondering whether we should not log the FullQualifiedUsername due to possible PII concerns. WDYT? any suggestions?

Reply:
changed to `user.toFullQualifiedUsername()`